File: C_Alghorithms/C_Alghorithms/merge_sort.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <tuple>
#include <algorithm>


class CrossoverFinder 
{
    public:
        CrossoverFinder(const std::vector<int>& x, const std::vector<int>& y)
            : x(x), y(y) {
            check_input(x, y);
        }

        int find_crossover_index() {
            return find_crossover_index_helper(0, x.size() - 1);
        }

    private:

        std::vector<int> x;
        std::vector<int> y;

        void check_input(const std::vector<int>& x, const std::vector<int>& y) {
            if (x.size() != y.size()) {
                std::cerr << "x and y must have the same length" << std::endl;
                exit(EXIT_FAILURE);
            }
            if (x[0] <= y[0]) {
                std::cerr << "x[0] must be greater than y[0]" << std::endl;
                exit(EXIT_FAILURE);
            }
            if (x.back() >= y.back()) {
                std::cerr << "x[-1] must be less than y[-1]" << std::endl;
                exit(EXIT_FAILURE);
            }
        }

    int find_crossover_index_helper(int left, int right) {
        int mid = (left + right) / 2;

        if (x[mid] > y[mid] && x[mid + 1] < y[mid + 1]) {
            return mid;
        }
        else if (x[mid] > y[mid]) {
            return find_crossover_index_helper(mid + 1, right);
        }
        else {
            return find_crossover_index_helper(left, mid);
        }
    }
};


class TestCrossoverFinder 
{
    public:
        static void run_tests() {

            std::vector<std::tuple<std::vector<int>, std::vector<int>, std::vector<int>>> test_cases = {
                std::make_tuple(std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7}, std::vector<int>{-2, 0, 4, 5, 6, 7, 8, 9}, std::vector<int>{1}),
                std::make_tuple(std::vector<int>{0, 1, 2, 3, 4, 5, 6, 7}, std::vector<int>{-2, 0, 4, 4, 4, 4, 8, 9}, std::vector<int>{1, 5}),
                std::make_tuple(std::vector<int>{0, 1}, std::vector<int>{-10, 10}, std::vector<int>{0}),
                std::make_tuple(std::vector<int>{0, 1, 2, 3}, std::vector<int>{-10, -9, -8, 5}, std::vector<int>{2})
            };

            for (size_t i = 0; i < test_cases.size(); ++i) {

                const std::vector<int>& x = std::get<0>(test_cases[i]);
                const std::vector<int>& y = std::get<1>(test_cases[i]);
                const std::vector<int>& expected = std::get<2>(test_cases[i]);

                CrossoverFinder finder(x, y);
                int result = finder.find_crossover_index();

                // Check if result is in the list of expected results
                if (std::find(expected.begin(), expected.end(), result) == expected.end()) {
                    std::cerr << "Test Case #" << i + 1 << " Failed: Expected one of {";
                    for (int val : expected) std::cerr << val << " ";
                    std::cerr << "} but got " << result << std::endl;
                    exit(EXIT_FAILURE);
                }
            }

            std::cout << "Mege Sort: All test cases passed." << std::endl;
        }
    };

File: C_Alghorithms/python/sorting_algorithms/merge_sort.py
# Merge sort algorithm
# Based on Coursera Algorithms for searching, sorting and Indexing


class CrossoverFinder(object):
    def __init__(self, x, y):

        self.x = x
        self.y = y
        self.check_input(x, y)

    @staticmethod
    def check_input(x, y):
        if not len(x) == len(y):
            print("x and y must have the same length")
        if not x[0] > y[0]:
            print("x[0] must be greater than y[0]")
        if not x[-1] < y[-1]:
            print("x[-1] must be less than y[-1]")

    def find_crossover_index(self):
        return self._find_crossover_index_helper(0, len(self.x) - 1)

    def _find_crossover_index_helper(self, left, right):
        mid = (left + right) // 2

        if self.x[mid] > self.y[mid] and self.x[mid + 1] < self.y[mid + 1]:
            return mid
        elif self.x[mid] > self.y[mid]:
            return self._find_crossover_index_helper(mid + 1, right)
        else:
            return self._find_crossover_index_helper(left, mid)


class TestCrossoverFinder(object):

    @staticmethod
    def run_tests():
        test_cases = [
            ([0, 1, 2, 3, 4, 5, 6, 7], [-2, 0, 4, 5, 6, 7, 8, 9], 1),
            ([0, 1, 2, 3, 4, 5, 6, 7], [-2, 0, 4, 4.2, 4.3, 4.5, 8, 9], [1, 5]),
            ([0, 1], [-10, 10], 0),
            ([0, 1, 2, 3], [-10, -9, -8, 5], 2),
        ]

        for idx, (x, y, expected) in enumerate(test_cases, start=1):
            finder = CrossoverFinder(x, y)
            result = finder.find_crossover_index()
            if isinstance(expected, list):
                assert result in expected, f"Test Case #{idx} Failed"
            else:
                assert result == expected, f"Test Case #{idx} Failed"

        print('Merge Sort: All test cases passed')


if __name__ == "__main__":
    TestCrossoverFinder.run_tests()
